README
Micro Lua DS Simulator 0.5 (by Ced-le-pingouin)
===============================================
Note: I sometimes use ML for Micro Lua DS in the text below, or refer to my ML 
      simulator script as "the sim" or MLS


Many things happened in 2010, between MLS v0.4 and v0.5, and I would like to
dedicate this release to the following people:

  - Peter "The Green Man" Steele (1962-2010):
        Without you, music will be a lot more boring and dull. You will be 
        deeply missed, Giant.
  
  - Ronnie James Dio (1942-2010):
        Well, it seems the evil eye finally got you. Wherever you are now, 
        Heaven or Hell, rest assured that your work in heavy metal will never 
        be forgotten \m/
        
  - Nergal (from the band Behemoth):
        In this terrible year 2010, bad news kept coming, one of them being 
        your illness. Luckily, it seems you have now beaten that fuckin' 
        disease, and I hope it's for good. Now come back with an even more kick
        ass album than Evangelion. Hard to believe it's possible, but I know 
        you and your bandmates can do it. That's all I wish for you.
        
   ...and, last but by no means least:
   
   - Some mysterious (Raving? Nooooooo!) Rabbid:
        Despite all the bad news, you made 2010 one of the best years of my 
        life. Thank you, you are amazing.


---------------------
[0] Table of contents
---------------------
[1] What's this thing ?
[2] What's in the zip file ?
[3] Controls
[4] Config file
[5] Remarks, what's done and what's not, limitations
[6] The future
[7] Contact
[8] Thanks & links


-----------------------
[1] What's this thing ?
-----------------------
The short description is: a Lua script that allows you to run other Lua scripts
normally targeted at Micro Lua DS (i.e. using the stylus, the screens, 
timers...), directly on your computer (Linux, Windows, and Mac OS X).

If you're interested in the long story, keep reading. 

I recently discovered Micro Lua DS, an interpreter for the Lua language on the 
Nintendo DS, fully compatible with Lua 5.1, with additional objects and methods
to use the DS controls, screens, sounds, sprites...

I started to think about writing some small games for the DS in Micro Lua, as I
found the setup of devkitARM and my return to "unmanaged" C programming too
time-consuming and tedious for the time I have right now.

But I'm running Linux (Ubuntu), and the available NDS emulators all have some 
problems on my machine, be it the lack of DLDI support that Micro Lua needs, or 
some segfaults I cannot explain.

So the only solution I had was to run a *windows* (!!!) emulator (iDeaS) under 
Wine, but even on a fast machine, there seems to be important lag in the display
when used this way.

Another possibility would be to transfer my Lua scripts on my DS on every 
change, but it's boring.

I could also edit my scripts directly on the DS using the excellent Lua Editor 
or CodeMonkeyDS, but let's face it, I type so much slower in these than on my PC
keyboard... ;)

So in a crazy moment, I decided to write some sort of Micro Lua simulator for 
"standard" Lua, i.e. I wanted to be able to develop directly in Lua on my PC, 
using the additional objects provided by Micro Lua for the DS, and I wanted my 
programs to be displayed on the PC as they would on the DS.

To achieve this, I had to find a way to make the non-standard Micro Lua objects 
and methods available on standard Lua under Linux. I took a look at what libs 
and bindings were available for Lua, supporting windowing, keyboard and mouse
inputs, image manipulation, timers, and maybe sounds. Since I wanted to begin 
quickly and viewed this project as an exercise for beginning in Lua, I chose 
wxWidgets for my main lib, the only thing missing being the sound (well, there's
wxSound, but it only plays WAV files).

I quickly released my first, very imperfect attempt at this project, lacking 
sound and being far from optimized, with some missing or incorrect behaviours 
compared to the real Micro Lua DS.

I intend to add sound in future versions (probably using SDL bindings), as well 
as optimize what can be, and maybe later use something else than wxWidgets, 
which is probably not that fast for bitmap and image manipulations. I could look
into Qt bindings, or try to use an OpenGL canvas for the graphics, but I have 
close to no experience of OpenGL, so I guess this would be the last thing I try.

PS: since the initial text, OpenGL has been implemented, yay!


----------------------------
[2] What's in the zip file ?
----------------------------
The main script mls.comp.lua, the Lua 5.1 executable, some libs, some images, 
and a "mls" launcher script.

That is if you have the "compiled" version of MLS. Starting with v0.4, MLS is 
free software, so you can have a "source" version. In this case, you'll also 
have a hierarchy of source classes and packages, some maintenance scripts (only
for Linux right now), and an additional "mls-src" launcher script, which starts 
the non-compiled version of MLS. This means that if you change the source and 
want to see what happens, you must use this launcher, since the other one 
launches the "compiled" version.

Please note that the "source" version is rather raw, and that it includes the
binaries and launching scripts for the three operating systems. As a result, 
those binaries and scripts aren't by default at the root of the archive, and you
should copy them from one of the subdirectories of "os_specific", depending on 
which OS you use. These subfolders are obviously named "linux", "mac", and 
"win".

You can run your Micro Lua script by typing the following command at the prompt,
while in the MLS directory:

Linux & Mac OS X: ./mls <name-of-your-script>
Windows: mls <name-of-your-script>

(don't forget to use mls-src instead if you want to run the source version!)

The script name is now optional, as MLS will launch a "boot script" if you don't
give it a user script. See config file for details (boot_script option).

IMPORTANT: the mls or mls-src script should be started with its directory as 
the current directory. If it's started from elsewhere (e.g. it's in your PATH so
it can be started from anywhere), MLS won't find some of its files (the config 
file, and some images).


------------
[3] Controls
------------
Of course the stylus is replaced by the mouse and its pression by the left 
mouse button.

In this version, the buttons<>keys bindings cannot be changed. They are:

+-----------+-----------------------------+
|    DS     |             Key             |
+-----------+-----------------------------+
| Pad Up    | I, Up arrow or KP 8         |
| Pad Down  | K, Down arrow, KP 5 or KP 2 |
| Pad Left  | J, Left arrow or KP 4       |
| Pad Right | L, Right arrow or KP 6      |
| L         | A or R                      |
| R         | Z or T                      |
| A         | D                           |
| B         | X                           |
| X         | E                           |
| Y         | S                           |
| Start     | Q or F                      |
| Select    | W or V                      |
+-----------+-----------------------------+

Starting with v0.4, additional keys and combinations not related to the DS are 
available:

+------------------------------------+----------+
|               Action               |    Key   |
+------------------------------------+----------+
| Open a script                      | Ctrl+O   |
| Pause script                       | P        |
| Restart script                     | B        |
| Reload script (from disk)          | Shift+B  |
| Show/hide console                  | C        |
| Clear console                      | DEL      |
| Decrease FPS                       | F1       |
| Increase FPS                       | F2       |
| Decrease UPS                       | F3       |
| Increase UPS                       | F4       |
| Simple/advanced drawGradientRect() | F5       |
| Adjust rectangles length           | F6       |
| Switch between 1x and 2x zoom      | F7       |
| Stylus.newPress hack (ML3/ML2)     | H        |
| Fullscreen mode (Windows only)     | F11      |
| Increase log level                 | F12      |
| Show key bindings summary          | Ctrl+K   |
| Quit MLS                           | Ctrl+Q   |
+------------------------------------+----------+

The Ctrl key is replaced by Cmd (="apple") on Mac OS X.

The console is where error messages from running script will be displayed.
It's also there that warning log messages from MLS will appear.

The FPS are the frames/sec that MLS will try to achieve, these are *display* 
frames and only affect the smoothness of the display. The default value is 60, 
even though ML claims a 30fps display rate. 60 maybe easier on the eye, anyway 
it should not change the *running* speed of the scripts (see below).

The UPS are updates/sec, that is the number of times MLS will iterate the main 
loop of a running script. I consider calls to Controls.read() in scripts as the
end of a loop iteration, since it is a common structure for game scripts.
So, these UPS will affect the real speed of the scripts. I don't know the exact
rate of the real ML, but the default value is 60.

Please note that "target" FPS and UPS won't be perfectly achieved, since my 
timing system is far from perfect, and depends on OS internals.

The log level (F12 key) is only there for debugging purposes, and you should not
play with this unless you know what you're doing. Otherwise the console might
overflow with useless MLS internal messages, and maybe slow down the running 

CHANGES
0.5 beta 1 (6 April 2011)
-------------------------
Changes for users:
* added the fake_root option in the config file, which will allow emulating SD
  card root by replacing the initial "/" in absolute paths with the "fake root"
  directory on your machine. By default, it is ./sdcard in MLS directory
* as an addition to the above feature, "fat:/" paths are now supported (I guess
  they're mostly used in older scripts anyway)
* it is now possible to disable "libs emulation", i.e. remove the libs and 
  functions made available by the default ML shell/libs. So you can test the 
  shell or write your own and test it with MLS
* it is now possible to choose a boot script (=shell) that can automatically
  start if you don't provide a script name on the command line. By default, it
  is /lua/libs/libs.lua, just like in ML, but you'll have to uncomment it in
  the config file, and copy the shell/libs file in the right location. If you 
  use the "fake root" described above, you should put the libs in 
  <fake root>/lua/libs/libs.lua (=./sdcard/lua/libs/libs.lua if you haven't 
  changed the default fake root)
* Sprite: start/stop/resetAnimation() are implemented (finally!)
* Image: mirrorH() and mirrorV() behavior was incorrect
* Canvas: X2 and Y2 handling was wrong in most functions
* OpenGL: screen.drawGradientRect() clipping was wrong
* OpenGL: screen.drawTextBox() clipping was buggy
* OpenGL: fixed many transparency issues
* OpenGL: Canvas.newPoint() no longer crashes
* OpenGL: ScrollMap is faster (when setTile() is not used, at least)
* wx: ScrollMap.draw() didn't repeat correctly 
* wx: Canvas.newPoint() didn't work on the lower screen
* wx: screen.drawGradientRect() didn't work when x0 > x1 or y0 > y1
* wx: screen.drawGradientRect() crashed with some Color parameters
* System: listDirectory() now puts "." and ".." before other items
* reloading a script from disk would sometimes crash
* log and error messages containing long paths are more readable
* boolean options in config were sometimes read incorrectly

Changes for the developer:
* Libs emulation facts:
  - some modules have a different name when libs emulation is disabled: 
    Controls => ds_controls, System => ds_system
  - Timer, Debug, DateTime, Sprite and INI are not defined when libs emulation 
    is disabled (in ML, they're defined by the default shell/libs)
  - some "internal" functions, undocumented for users, are available (in screen,
    ds_system, ds_controls...)
* Class: inheritance support is faster. And comes in two versions: "local" 
  (the default) and "global" classes
* Canvas: ATTR_ constants have the same values as in ML
* screen: SCREEN_UP and SCREEN_DOWN constants have the same values as in ML
* compile scripts: new version written in Lua (make and package don't use it 
  yet, though)
* compile scripts: it is now possible to keep debug symbols when compiling with
  Lua AIO
* compile scripts: class names in compiled version are less likely to clash with
  global variables in user scripts

0.4 (2 March 2010)
------------------
  ! OpenGL is now the default display mode in mls.ini
  * OpenGL fixes: no more white rectangles instead of images, or garbled images,
    or unbearably low FPS (on older GPUs that switched to software rendering)
    If you still have problems with OpenGL, notably with Pause, take a look at
    mls.ini, there are additional options for OpenGL
  * screen.drawTextBox() now supports newlines
  * new "key bindings" dialog in Help menu (or Ctrl+K)
  * zoom: fix cases where the zoom factor wouldn't switch back to 1x
  * zoom: disable zoom change while in fullscreen mode
  * mls.ini config file is provided instead of mls.ini.dist (which had to be 
    renamed by the user)

0.4 beta 3 (21 February 2010)
-----------------------------
  * OpenGL support. It can be enabled in the config file
  * the main window can now be resized, and a shortcut key (F7) allows you to 
    switch between 1x and 2x zoom. On Windows, F11 enables/disables fullscreen 
    (a bit ugly because of the non integer zoom ratio).
    WARNING: resizing the window can cause massive slowdown if OpenGL is 
    disabled (except maybe on Windows)
  * it is now possible to reload a script from disk (Shift+B)
  * initial current working directory is now the same as the running script
  * shortcut keys: increasing/decreasing FPS/UPS now uses steps of 5, instead of
    1
  * shortcut keys: displaying and hiding the console is now bound to F12 instead
    of F5
  * arrow keys now work on Windows too
  * screen: drawGradientRect() accuracy can now be adjusted (F5 key and config 
    file option). This option does nothing in OpenGL mode
  * screen: it is now possible to enable or disable "rectangle length 

clp/mls/Mls.lua
-------------------------------------------------------------------------------
-- The main class that should be instantiated for Micro Lua Simulator to start.
--
-- @class module
-- @name clp.mls.Mls
-- @author Ced-le-pingouin <Ced.le.pingouin@gmail.com>
--
-- @todo The spr_depl script (maybe others?) sometimes crashes suddenly, often 
--       after other scripts having been loaded before. I'm suspecting the many 
--       Images needed for Font cache. Maybe some <module>:resetModule() should 
--       dispose of resources (and thus keep track of the resources they load
--       or create)
-- @todo In aloufs, the + and - buttons are always clicked twice !?
-- @todo In UFO, under Linux, the click doesn't work well when L is held (screen
--       goes black). Is this related to the "click twice" problem in aloufs ?
-- @todo Some scripts are less responsive since the changes in timing (aloufs)
-- @todo Some scripts don't work: CodeMonkeyDS, LED 1.2b (in files menu when
--       pressed Back). alternativ-keyboard is dog-slow, too
-- @todo Have proper packaging for MacOS (as a real App)
--
-- @todo INI, Map, Mod, Sound, Sprite, ScrollMap: raise an error when a file is
--       not found ? Check what ML does when it happens. I've already done this
--       for Image.load()
-- @todo Choose which ML version is simulated (2.0/3.0) by (un)loading some 
--       modules and deleting some vars/constants (for ML 2).
--       Maybe allow additional boolean "hack" for these cases:
--           - ML 2 had a different behaviour for Stylus.newPress. See 
--             unmodified StylusBox lib with Stylus.newPressinBox() working in 
--             ML 2 but not in ML 3 (DS-laby and PPC DS use this lib)
--             In ML2, newPress is always true, except when held is true. So it
--             probably is "not held". In ML3, newPress occurs only once, when
--             the stylus has just been pressed. After that it is false again
--           - Command-EZ / CommandButton (L49): screen.drawRect() will crash in
--             MLS, see comments in that function for a working hack
--
-- @todo Have a test directory
-- @todo Refactor/split some classes (split ScriptManager/Script? split 
--       Gui/Console? Sys? Map/ScrollMap ?)
-- @todo Get key bindings from the ini file, too
-- @todo Have menu items for every config option
-- @todo Save config on exit
-- @todo Remember last directory we loaded a script from
-- @todo Remember recently loaded scripts
-- @todo Toolbar with shortcut buttons (open file, pause/resume...)
-- @todo Delete calls to Logger in compiled version ?
-- @todo Succeed in running LuaDoc on the source code
-- @todo Make the compilation and packaging script work in Windows (Mac?)
--
-- @todo Simulate real ML DS limits, e.g. on the number of Images that can be 
--       loaded, count used memory in RAM/VRAM...
-- @todo In all modules, search for temporary wx objects created on the fly 
--       like brushes, colors, points, pens... that are often used (e.g. white 
--       pen, point(0,0)) and see if I can make them pre-created objects, so 
--       there's no need to re-create them all the time
-------------------------------------------------------------------------------

--  Copyright (C) 2009-2011 Cédric FLOQUET
--
--  This file is part of Micro Lua DS Simulator.
--
--  Micro Lua DS Simulator is free software: you can redistribute it and/or modify
--  it under the terms of the GNU General Public License as published by
--  the Free Software Foundation, either version 3 of the License, or
--  (at your option) any later version.
--
--  Micro Lua DS Simulator is distributed in the hope that it will be useful,
--  but WITHOUT ANY WARRANTY; without even the implied warranty of
--  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
--  GNU General Public License for more details.
--
--  You should have received a copy of the GNU General Public License
--  along with Micro Lua DS Simulator.  If not, see <http://www.gnu.org/licenses/>.


require "wx"
local Class = require "clp.Class"
local Observable = require "clp.Observable"
local Logger = require "clp.Logger"
local Sys = require "clp.mls.Sys"
local Config = require "clp.mls.Config"
local Gui = require "clp.mls.Gui"
local DebugWindow = require "clp.mls.DebugWindow"
local Timer = require "clp.mls.modules.wx.Timer"
local ModuleManager = require "clp.mls.ModuleManager"
local ScriptManager = require "clp.mls.ScriptManager"
local Dispatcher = require "clp.mls.Dispatcher"

Mls = Class.new(Observable)

Mls.VERSION = "0.5beta1"

--- Constructor.
--
-- Creates and initializes the app main window, and the ML simulated modules
--
-- @param scriptPath (string) The path of an initial script to run
function Mls:ctr(scriptPath)
    Mls.logger = Logger:new(Logger.WARN, "*")
    
    Mls.initialDirectory = wx.wxGetCwd()
    
    -- two config files are valid, first the dev one, then the user one
    local configFile, found = nil, false
    for _, possibleConfigFile in ipairs{ "mls.dev.ini", "mls.ini" } do
        configFile, found = Sys.getFile(possibleConfigFile)
        if found then break end
    end
    if not found then configFile = nil end
    Mls.config = Config:new(configFile, "mls", Mls:getValidOptions())
    
    Mls.logger:setLevel(Mls.config:get("debug_log_level", Logger.WARN))
    
    
    -- a boot script is launched if no command-line script is given
    local bootScript = Mls.config:get("boot_script", "/lua/libs/libs.lua")
    if scriptPath then
        bootScript = nil
    else
        scriptPath = bootScript
    end
    
    -- should MLS emulate libs? (the default is no if a boot script is used)
    local emulateLibs = not bootScript
    emulateLibs = Mls.config:get("emulate_libs", emulateLibs)
    
    -- set "fake root"
    local fakeRootDefault = Sys.buildPath(Mls.initialDirectory, "sdcard")
    Sys.setFakeRoot(Mls.config:get("fake_root", fakeRootDefault))
    
    
    -- init vars and gui
    Mls._initVars()
    Mls.keyBindings = Mls._loadKeyBindingsFromFile("README")
    Mls.gui = Mls._initGui(Mls.initialDirectory)
    
    -- debug window
    --[[
    local debugWindow = DebugWindow:new()
    debugWindow:show()
    debugWindow:setSourceFile("./clp/mls/Mls.lua")
    debugWindow:setCurrentLineInSource(384)
    debugWindow:setCurrentLineInSource(386)
    debugWindow:setGridVariables(_G)
    debugWindow:sortGridByColumn(0)
    ]]
    
    -- logger
    Mls.logger:setWriterFunction(Mls.gui:getConsoleWriter())
    Mls.logger:setLogFormat("%m")
              :reserved("Welcome to the console. Script errors and log messages will be displayed here.")
              :resetLogFormat()
    
    -- info on the Class system used for inheritance and late static binding
    local globalClassesState = Class.globalClassesEnabled()
                               and "global classes"
                                or "local classes"
    
    Mls.logger:info("Class is set up for "..globalClassesState)
    
    -- debug vars
    __DEBUG_NO_REFRESH = Mls.config:get("debug_no_refresh", false)
    __DEBUG_LIMIT_TIME = Mls.config:get("debug_limit_time", 0)
    
    -- OpenGL stuff
    Mls.openGl = Mls.config:get("open_gl", false)
    Mls.openGlUseTextureRectangle = Mls.config:get(
        "open_gl_use_texture_rectangle", true
    )
    Mls.openGlSimplePause = Mls.config:get("open_gl_simple_pause", false)
    
    -- ML modules manager
    local moduleManager = ModuleManager:new()
    if Mls.openGl then
        moduleManager:addPrefix("gl.", true)
    end
    
    moduleManager:enableLibsEmulation(emulateLibs)
    
    -- script manager
    local fps = Mls.config:get("fps", 60)
    local ups = Mls.config:get("ups", 60)
    local timing = Mls.config:get("debug_main_loop_timing", nil)
    
    Mls.scriptManager = ScriptManager:new(fps, ups, timing, moduleManager)
    
    -- event handlers
    Mls:attach(self, "scriptStateChange", self.onScriptStateChange)
    Mls:attach(self, "upsUpdate", self.onUpsUpdate)
    Mls:attach(self, "keyDown", self.onKeyDown)
    if __DEBUG_LIMIT_TIME > 0 then
        Mls:attach(self, "stopDrawing", self.onStopDrawing)
    end
    
    -- timer
    Mls._initTimer()
    
    -- hacks and config options
    ds_controls.setStylusHack(Mls.config:get("stylus_hack", false))
    screen.setDrawGradientRectAccuracy(
        Mls.config:get("draw_gradient_rect_accuracy", 0)
    )
    screen.setRectAdditionalLength(Mls.config:get("rect_length", 0))
    
    -- MLS dispatcher
    local frank = Dispatcher:new()
    frank:dispatch():enableItemFetching()
    
    --print(table.concat(frank:encodeData(""), ", "))
    --print(frank:decodeData({}))
    
    -- and finally load the script given at the command line if needed, or the
    -- "boot script" that is defined in the config file
    Mls.scriptManager:init()
    
    if scriptPath then
        Mls.scriptManager:loadAndStartScript(Sys.getFile(scriptPath))
    end
    
    -- in case some module has changed GUI components, we re-set the focus
    Mls.gui:focus()
end

--- Initializes ML global and internal variables.
function Mls._initVars()
    Mls.logger:info("initializing variables")
    
    MICROLUA_VERSION = "3.0"
    
    SCREEN_WIDTH  = 256
    SCREEN_HEIGHT = 192
    Mls.DEPTH = -1
    
    Mls.description = ""
end

--- Initializes main window, menu items and their associated action, then shows
--  the window.
--
-- @param path (string) An optional directory to search for the GUI icons
--
-- @return (Gui) The created Gui object
function Mls._initGui(path)
    Mls.logger:info("initializing GUI")
    
    local gui = Gui:new(SCREEN_WIDTH, SCREEN_HEIGHT * 2, 
                        "MLS "..Mls.VERSION, nil, path)
    
    gui:registerShutdownCallback(function()
        -- the events functions must always be called with current "Mls"
        Mls.displayInfo()
        gui:shutdown()
        collectgarbage("collect")
        os.exit(0)
    end)
    
    gui:createMenus{
        {
            caption = "&File",
            items   = {
                {
                    caption  = "&Open...",
                    id       = gui.MENU_OPEN,
                    callback = Mls.onFileOpen
                },
                {
                    caption  = "E&xit",
                    id       = gui.MENU_EXIT,
                    callback = Mls.onExit
                }
            }
        },
        {
            caption = "&Help",
            items   = {
                {
                    caption  = "&About",
                    id       = gui.MENU_ABOUT,
                    callback = Mls.onAbout
                },
                {
                    caption  = "Show &key bindings",
                    id       = gui.MENU_SHOW_KEY_BINDINGS,
                    callback = Mls.onShowKeyBindings
                }
            }
        }
    }
    
    gui:showWindow()
    
    return gui
end

--- Initializes the main lib timer.
function Mls._initTimer()
    Mls.logger:info("initializing timer")
    
    Mls._timer = Timer.new()
    Mls._timer:start()
    Mls._startTime = Mls._timer:time()
end

--- Reads key bindings from a file.
--
-- The file should contains key bindings as the first two "tables", as in the 
-- README. That is, a table begins with "+-----" then two lines for headers, and
-- ends with "+-----". Everything between the header and the end of the "table"
-- is seen as an "action - key binding" pair. Columns are drawn with "|", so 
-- there are three "|" for each line of the "table".
-- e.g. "|   Pause/resume   |    P   |"
--
-- @param fileName (string)
--
-- @return (array) Each item will itself be an array, where the first item is a
--                 string that describes the action, and the second item is a
--                 string that represents the key associated to it.
--                 e.g. { { "Pause/resume", "P" }, { "Quit", "Ctrl+Q" } }
function Mls._loadKeyBindingsFromFile(fileName)
    local isMac = (Sys.getOS() == "Macintosh")
    local keyBindings = {}
    
    local file = io.open(fileName, "r")
    
    local maxTables = 2
    local numTables = 0
    local inTable = false
    
    while true do
        local line = file:read()
        if not line then break end
        
        if line:find("^%+%-%-%-%-%-") then
            inTable = not inTable
            
            if inTable then
                file:read()
                file:read()
            else
                numTables = numTables + 1
                if numTables >= maxTables then break end
            end
        elseif inTable then
            local action, key = line:match("|%s+(.-)%s+|%s+(.-)%s+|")
            
            -- on Mac, Ctrl is not used for key combinations, Cmd/Apple is
            if isMac then key = key:gsub("Ctrl%+", "Cmd+") end
            
            table.insert(keyBindings, { action, key })
        end
    end
    
    file:close()
    
    return keyBindings
end

--- Returns the list of valid config options for MLS.
--
-- @return (table)
--
-- @see Config._validateOption to understand the format of an option
function Mls:getValidOptions()
    Mls.logger:info("reading allowed config options")
    
    return {
        fake_root = { "string" },
        boot_script = { "string" },
        emulate_libs = { "boolean" },
        fps = { "number", 0 },
        ups = { "number", 0 },
        bitmap_fonts = { "boolean" },
        stylus_hack = { "boolean" },
        rect_length = { "number", 0, 1 },
        draw_gradient_rect_accuracy = { "number", 0, 256 },
        open_gl = { "boolean" },
        open_gl_use_texture_rectangle = { "boolean" },
        open_gl_simple_pause = { "boolean" },
        
        -- debug options below
        debug_log_level = { "number", Logger.TRACE, Logger.FATAL },
        debug_main_loop_timing = { "number", { ScriptManager.TIMING_BUSY, 
                                               ScriptManager.TIMING_IDLE, 
                                               ScriptManager.TIMING_TIMER } 
        },
        debug_no_refresh = { "boolean" },
        debug_limit_time = { "number", 0 },
    }
end

--- Displays fps/ups info.
function Mls.displayInfo()
    local elapsedTime = (Mls._timer:time() - Mls._startTime) / Timer.ONE_SECOND
    local totalFps = screen.getUpdates() / elapsedTime
    local totalUps = Mls.scriptManager:getUpdates() / elapsedTime
    Mls.logger:info(string.format("%d secs - %d fps - %d ups", 
                                   elapsedTime, totalFps, totalUps))
end

--- Quits the app if the debug variable limiting execution time is set, and time
--  is over.
--
-- Called on stopDrawing events
--
-- @eventHandler
function Mls:onStopDrawing()
    if (self._timer:time() - self._startTime) > __DEBUG_LIMIT_TIME then
        self:delete()
    end
end

--- Handles keys that are not part of ML, such as FPS/UPS modification, log 
--  level modification, pause/resume script, reset script, show/hide GUI 
--  console, clear GUI console...
--
-- Called on keyDown events
--
-- @param event (string) The name of the event that caused the callback. 
--                       Should be "keyDown" here
-- @param key (number) The raw key code
-- @param shift (boolean) true if the Shift key is pressed
--
-- @eventHandler
function Mls:onKeyDown(event, key, shift)
    local fpsAndUpsStep = 5
    local sm = Mls.scriptManager
    
    if key == wx.WXK_P then
        if not shift then
            sm:pauseOrResumeScript()
        else
            sm:debugOneStepScript()
        end
    elseif key == wx.WXK_B then
        if not shift then
            sm:restartScript()
        else
            sm:reloadAndStartScript()
        end
    elseif key == wx.WXK_C then
        Mls.gui:showOrHideConsole()
    elseif key == wx.WXK_H then
        ds_controls.switchStylusHack()
    elseif key == wx.WXK_DELETE or key == wx.WXK_NUMPAD_DELETE then
        Mls.gui:clearConsole()
    elseif key == wx.WXK_F1 then
        sm:setTargetFps(sm:getTargetFps() - fpsAndUpsStep)
    elseif key == wx.WXK_F2 then
        sm:setTargetFps(sm:getTargetFps() + fpsAndUpsStep)
    elseif key == wx.WXK_F3 then
        sm:setTargetUps(sm:getTargetUps() - fpsAndUpsStep)
    elseif key == wx.WXK_F4 then
        sm:setTargetUps(sm:getTargetUps() + fpsAndUpsStep)
    elseif key == wx.WXK_F5 then
        screen.switchDrawGradientRectAccuracy()
    elseif key == wx.WXK_F6 then
        screen.incRectAdditionalLength()
    elseif key == wx.WXK_F7 then
        Mls.gui:incZoomFactor()
    elseif key == wx.WXK_F11 then
        Mls.gui:switchFullScreen()
    elseif key == wx.WXK_F12 then
        Mls.logger:incrementLevel(true)
    end
end

--- Displays ups and fps information in the Gui.
--
-- Called on upsUpdate events
--
-- @param event (string) The name of the event that caused the callback. 
--                       Should be "upsUpdate" here
-- @param ups (number) The current ups (main loop updates/sec = iterations/src)
--
-- @eventHandler
function Mls:onUpsUpdate(event, ups)
    local targetFps = Mls.scriptManager:getTargetFps()
    local targetUps = Mls.scriptManager:getTargetUps()
    
    Mls.gui:displayTimingInfo(string.format(
        "%d fps (%d) - %d ups (%d)", screen.getFps(), targetFps, ups, targetUps
    ))
end

--- Displays script name and state in the Gui, and shows GUI console when a
--  script error has occured.
--
-- Called on scriptStateChange events
--
-- @param event (string) The name of the event that caused the callback. 
--                       Should be "scriptStateChange" here
-- @param script (string) The name of the script that changed state
-- @param state (number) The new state of the script (ScriptManager SCRIPT_ 
--                       constants)
--
-- @eventHandler
function Mls:onScriptStateChange(event, script, state)
    self.gui:displayScriptName(script)
    self.gui:displayScriptState(ScriptManager.getStateName(state))
    
    if state ~= ScriptManager.SCRIPT_RUNNING then
        local color
        if state == ScriptManager.SCRIPT_NONE 
           or state == ScriptManager.SCRIPT_ERROR
        then
            color = Color.new(31, 0, 0)
        elseif state == ScriptManager.SCRIPT_PAUSED then
            -- "paused" banner = color of OpenGL logo in OpenGL, green otherwise
            color = Mls.openGl and Color.new(11, 17, 21) or Color.new(0, 20, 0)
        else
            color = Color.new(0, 0, 31)
        end
        
        local sm = Mls.scriptManager
        local caption = nil
        if not sm:debugModeEnabled() then
            caption = sm.getStateName(state):upper()
        end
        
        screen.displayInfoText(caption, color)
    end
    
    if state == ScriptManager.SCRIPT_NONE
       or state == ScriptManager.SCRIPT_ERROR
    then
        self.gui:showOrHideConsole(true)
    end
end

--- Opens a script file selection dialog and runs the chosen script.
function Mls.onFileOpen()
    Mls.logger:debug("begin Open", "menu")
    
    Mls.scriptManager:pauseScriptWhile(function()
        local file = Mls.gui:selectFile{
            caption     = "MLS - Select a Lua script to run",
            defaultPath = "",
            defaultFile = "",
            defaultExt  = "lua",
            filters = {
                ["*.lua"] = "Lua scripts (*.lua)",
                ["*.*"] = "All files (*.*)",
            }
        }
        
        if file ~= "" then
            screen.clearAllOffscreenSurfaces()
            Mls.scriptManager:loadAndStartScript(file)
        end
    end)
    
    Mls.gui:focus()
    
    Mls.logger:debug("end Open", "menu")
end

--- Shows the application About box.
function Mls.onAbout()
    Mls.logger:debug("begin About", "menu")
    
    Mls.scriptManager:pauseScriptWhile(function()
        Mls.gui:showAboutBox{
            name = "Micro Lua Simulator (aka MLS)",
            version = Mls.VERSION,
            description = "Run Micro Lua DS scripts on your computer"
                          .."\n\n"..Mls.description,
            authors = { "Ced-le-pingouin <Ced.le.pingouin@gmail.com>" },
            copyright = "(c) 2009-2011 Ced-le-pingouin",
            link = {
                caption = "Micro Lua Simulator topic on official Micro Lua DS forums",
                url = "http://microlua.xooit.fr/t180-Micro-Lua-Simulator.htm"
            },
            license = [[
 Micro Lua DS Simulator is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 Micro Lua DS Simulator is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with Micro Lua DS Simulator.  If not, see <http://www.gnu.org/licenses/>.
            ]]
        }
    end)
    
    Mls.logger:debug("end About", "menu")
end

--- Shows the key bindings
function Mls.onShowKeyBindings()
    Mls.gui:showKeyBindings(Mls.keyBindings)
end

--- Requests the application to close.
--
-- Sends a request to the GUI, the app will only exit after the main window has 
-- closed
function Mls.onExit()
    Mls.logger:debug("Exit", "menu")
    
    Mls.gui:closeWindow()
end

--- Releases resources used by MLS and requests the app main frame to close.
function Mls.delete()
    Mls._frame:Disconnect(wx.wxEVT_TIMER)
    Mls._frame:Close()
    wx.wxYield()
end
